src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,332 @@
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_connect(b_obj_arg socket, b_obj_arg addr) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    sockaddr_in addr_ptr;
+    lean_socket_addr_to_sockaddr(addr, &addr_ptr);
+
+    lean_object * promise = create_promise();
+
+    uv_connect_t * uv_connect = (uv_connect_t*)malloc(sizeof(uv_connect_t));
+    uv_connect->data = promise;
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_tcp_connect(uv_connect, tcp_socket->m_uv_tcp, (const struct sockaddr *)&addr_ptr, [](uv_connect_t* req, int status) {
+        lean_object * promise = (lean_object *) req->data;
+        resolve_promise(promise, status);
+        free(req);
+    });
+
+    if (result < 0) {
+        free(uv_connect);
+        return lean_io_result_mk_error(mk_except_error_string("Connection failed"));
+    }
+
+    event_loop_unlock(&global_ev);
+
+    return lean_io_result_mk_ok(promise);
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_send(b_obj_arg socket, b_obj_arg data) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    size_t data_len = lean_sarray_size(data);
+    char * data_str = (char *)lean_sarray_cptr(data);
+    uv_buf_t buf = uv_buf_init(data_str, data_len);
+
+    lean_object * promise = create_promise();
+
+    event_loop_lock(&global_ev);
+
+    uv_write_t * write_uv = (uv_write_t*)malloc(sizeof(uv_write_t));
+    write_uv->data = promise;
+
+    int result = uv_write(write_uv, (uv_stream_t*)tcp_socket->m_uv_tcp, &buf, 1, [](uv_write_t * req, int status) {
+        lean_object * promise = (lean_object *) req->data;
+        resolve_promise(promise, status);
+        free(req);
+    });
+
+    if (result < 0) {
+        free(write_uv);
+        return lean_io_result_mk_error(mk_except_error_string("Write failed"));
+    }
+
+    event_loop_unlock(&global_ev);
+    return lean_io_result_mk_ok(promise);
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_recv(b_obj_arg socket) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+    lean_object * promise = create_promise();
+
+    tcp_socket->m_promise_read = promise;
+    lean_inc(promise);
+
+    event_loop_lock(&global_ev);
+
+    uv_read_start((uv_stream_t*)tcp_socket->m_uv_tcp, [](uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf) {
+        buf->base = (char*)malloc(suggested_size);
+        buf->len = suggested_size;
+    }, [](uv_stream_t *stream, ssize_t nread, const uv_buf_t *buf) {
+        lean_uv_tcp_socket_object *tcp_socket = (lean_uv_tcp_socket_object *)stream->data;
+        lean_object * promise = tcp_socket->m_promise_read;
+
+        if (nread > 0) {
+            lean_object * data = lean_alloc_sarray(1, nread, nread);
+            memcpy(lean_sarray_cptr(data), buf->base, nread);
+            resolve_promise(promise, 0);
+            lean_dec(data);
+        } else if (nread < 0) {
+            resolve_promise(promise, nread);
+        }
+
```

Comment:
I would have expected the promise to be removed from our `struct`/cleared to `NULL` again here and it's ref count decremented such that we free it.